users_ldap_populate/models/users_ldap.py

```diff
@@ -0,0 +1,194 @@
+# Copyright 2021 Tecnativa - João Marques
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/gpl.html).
+
+import logging
+import re
+
+from odoo import SUPERUSER_ID, _, fields, models
+from odoo.exceptions import AccessDenied, UserError
+
+_logger = logging.getLogger(__name__)
+
+try:
+    import ldap
+    from ldap.filter import filter_format
+except ImportError:
+    _logger.debug("Cannot import ldap.")
+
+
+class CompanyLDAP(models.Model):
+    _inherit = "res.company.ldap"
+
+    no_deactivate_user_ids = fields.Many2many(
+        comodel_name="res.users",
+        relation="res_company_ldap_no_deactivate_user_rel",
+        column1="ldap_id",
+        column2="user_id",
+        string="Users never to deactivate",
+        help="List users who never should be deactivated by the deactivation wizard",
+        default=lambda self: [(6, 0, [SUPERUSER_ID])],
+    )
+    deactivate_unknown_users = fields.Boolean(
+        string="Deactivate unknown users", default=False,
+    )
+
+    def action_populate(self):
+        """
+        Prepopulate the user table from one or more LDAP resources.
+
+        Obviously, the option to create users must be toggled in
+        the LDAP configuration.
+
+        Return the number of users created (as far as we can tell).
+        """
+        _logger.debug("action_populate called on res.company.ldap ids %s", self.ids)
+        users_model = self.env["res.users"]
+        users_count_before = users_model.search_count([])
+        deactivate_unknown, known_user_ids = self._check_users()
+        if deactivate_unknown:
+            _logger.debug("will deactivate unknown users")
+        for conf in self._get_ldap_dicts():
+            if not conf["create_user"]:
+                continue
+            attribute_match = re.search(r"([a-zA-Z_]+)=\%s", conf["ldap_filter"])
+            if attribute_match:
+                login_attr = attribute_match.group(1)
+            else:
+                raise UserError(
+                    _(
+                        "No login attribute found: "
+                        "Could not extract login attribute from filter %s"
+                    )
+                    % conf["ldap_filter"]
+                )
+            results = self._get_ldap_entry_dicts(conf)
+            for result in results:
+                login = result[1][login_attr][0].lower().strip()
+                user_id = None
+                try:
+                    user_id = self.with_context(
+                        no_reset_password=True
+                    )._get_or_create_user(conf, login, result)
+                except AccessDenied:
+                    # this happens if the user exists but is active = False
+                    # -> fetch the user again and reactivate it
+                    self.env.cr.execute(
+                        "SELECT id FROM res_users WHERE lower(login)=%s", (login,)
+                    )
+                    res = self.env.cr.fetchone()
```

Comment:
IDK if this could be changed to improve security and prevent a potential SQLi risk. But then, it's also how Odoo does it: https://github.com/odoo/odoo/blob/13.0/addons/auth_ldap/models/res_company_ldap.py#L193
@Tardo @Yajo WDYT?

Reply:
I think it can be OK.